### PR TITLE
Atlas Cloud Watch: Add the PublishQueue class to replace the Client A…

### PR DESCRIPTION
**Pull request by manolama**
…ctor

grouping and flushing data points to Publish Proxy instances. A queue will be instantiated for each stack and a router (in the next PR) wil use the AWS account to route data points to the proper queue.

### CODE REVIEW

File: atlas-cloudwatch/src/main/scala/com/netflix/atlas/cloudwatch/PublishQueue.scala [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+
+  /**
+    * Adds the data point to the queue if there is room. Increments a metric if not.
+    *
+    * @param datapoint
+    *     The non-null data point to enqueue.
+    */
+  def enqueue(datapoint: AtlasDatapoint): Unit = {
+    if (!publishQueue.offer(datapoint)) {
+      droppedFull.increment()
```
Comment (brharrington) [inline atlas-cloudwatch/src/main/scala/com/netflix/atlas/cloudwatch/PublishQueue.scala:87]: There are standard metrics on the queue for reporting number of items submitted or dropped due to being full.

File: atlas-cloudwatch/src/main/scala/com/netflix/atlas/cloudwatch/PublishQueue.scala [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+        response.status.intValue() match {
+          case 200 => // All is well
+            response.discardEntityBytes()
+          case 202 | 206 => // Partial failure
+            val id = datapointsDropped.withTag("reason", "partialFailure")
+            incrementFailureCount(id, response, size)
+          case 400 => // Bad message, all data dropped
+            val id = datapointsDropped.withTag("reason", "completeFailure")
+            incrementFailureCount(id, response, size)
+          case 429 | 504 => // backoff
```
Comment (brharrington) [inline atlas-cloudwatch/src/main/scala/com/netflix/atlas/cloudwatch/PublishQueue.scala:118]: Was this supposed to be 503? We don't typically retry on timeouts because you cannot guarantee some of the data wasn't processed.

File: atlas-cloudwatch/src/main/scala/com/netflix/atlas/cloudwatch/PublishQueue.scala [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+            retry(payload, retries, size)
```
Comment (brharrington) [inline atlas-cloudwatch/src/main/scala/com/netflix/atlas/cloudwatch/PublishQueue.scala:119]: This could be done using [max-retries](https://doc.akka.io/docs/akka-http/current/client-side/host-level.html#retrying-a-request) setting on the client rather than explicitly here.

File: atlas-cloudwatch/src/main/scala/com/netflix/atlas/cloudwatch/PublishQueue.scala [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+  private val droppedRetries = registry.counter(datapointsDropped.withTags("reason", "maxRetries"))
+
+  private val maxRetries = getSetting("maxRetries")
+  private val queueSize = getSetting("queueSize")
+  private val batchSize = getSetting("batchSize")
+  private val batchTimeout = getDurationSetting("batchTimeout")
+
+  private[cloudwatch] val publishQueue = StreamOps
+    .blockingQueue[AtlasDatapoint](registry, s"${stack}PubQueue", queueSize)
+    .groupedWithin(batchSize, FiniteDuration.apply(batchTimeout.getSeconds, TimeUnit.SECONDS))
```
Comment (brharrington) [inline atlas-cloudwatch/src/main/scala/com/netflix/atlas/cloudwatch/PublishQueue.scala:74]: I would recommend doing:

```
FiniteDuration(batchTimeout.toNanos, TimeUnit.NANOSECONDS)
```

This avoids surprises if say we used a duration of 500ms.

File: atlas-cloudwatch/src/main/scala/com/netflix/atlas/cloudwatch/PublishQueue.scala [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+            retry(payload, retries, size)
+          case v => // Unexpected, assume all dropped
+            response.discardEntityBytes()
+            val id = datapointsDropped.withTag("reason", s"status_$v")
+            registry.counter(id).increment(size)
+        }
+
+      case Failure(ex) =>
+        logger.error(s"Failed publishing to ${uri} for ${stack}", ex)
+        incrementException(ex)
```
Comment (brharrington) [inline atlas-cloudwatch/src/main/scala/com/netflix/atlas/cloudwatch/PublishQueue.scala:128]: This is already covered by common IPC client metrics.

File: atlas-cloudwatch/src/main/scala/com/netflix/atlas/poller/Messages.scala [hunk 6]
```diff
@@ -42,7 +42,10 @@ object Messages {
     * @param metrics
     *     Metrics collected by the poller.
     */
-  case class MetricsPayload(tags: Map[String, String] = Map.empty, metrics: List[Datapoint] = Nil)
+  case class MetricsPayload(
+    tags: Map[String, String] = Map.empty,
+    metrics: Iterable[Datapoint] = Nil
```
Comment (brharrington) [inline atlas-cloudwatch/src/main/scala/com/netflix/atlas/poller/Messages.scala:47]: Why was this changed to `Iterable`?

File: atlas-cloudwatch/src/test/scala/com/netflix/atlas/cloudwatch/PublishQueueSuite.scala [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+    val queue =
+      new PublishQueue(config, registry, "main", "http://localhost", httpClient, scheduler)
+    queue.publish(
+      Seq(
+        Datapoint(Map("k1" -> "v1"), timestamp, 42.5),
+        Datapoint(Map("k2" -> "v2"), timestamp, 24.1)
+      )
+    )
+
+    Thread.sleep(threadSleep)
```
Comment (brharrington) [inline atlas-cloudwatch/src/test/scala/com/netflix/atlas/cloudwatch/PublishQueueSuite.scala:82]: Sleeps tend to make tests really flakey and should be avoided.

File: atlas-cloudwatch/src/test/scala/com/netflix/atlas/cloudwatch/PublishQueueSuite.scala [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+    droppedPartial: Long = 0,
+    droppedComplete: Long = 0,
+    dropped500: Long = 0,
+    droppedQueueFull: Long = 0,
+    droppedRetries: Long = 0,
+    exceptions: Long = 0
+  ): Unit = {
+    assertEquals(registry.counter("atlas.cloudwatch.queue.dps.sent", "stack", "main").count, sent)
+    assertEquals(registry.counter("atlas.cloudwatch.queue.retries", "stack", "main").count, retries)
+    assertEquals(
```
Comment (brharrington) [inline atlas-cloudwatch/src/test/scala/com/netflix/atlas/cloudwatch/PublishQueueSuite.scala:230]: This could be refactored to be less repetitive, maybe something like:

```scala
def assertCounters(
    sent: Long = 0,
    retries: Long = 0,
    exceptions: Long = 0
    dropped: Map[String, Long]
): Unit = {
  // check others
  List("partialFaliure", "completeFailure", "status_500", "queueFull", "maxRetries").foreach { reason =>
    assertEquals(
      registry
        .counter("atlas.cloudwatch.queue.dps.dropped", "stack", "main", "reason", reason)
        .count,
      dropped.getOrElse(reason, 0L)
    )
  }
}
```

File: atlas-cloudwatch/src/main/scala/com/netflix/atlas/cloudwatch/PublishQueue.scala (hunk 2 above)
Comment (manolama) [reply atlas-cloudwatch/src/main/scala/com/netflix/atlas/cloudwatch/PublishQueue.scala:118]: It was a 504 in the old pub proxy. I can drop it.

File: atlas-cloudwatch/src/main/scala/com/netflix/atlas/cloudwatch/PublishQueue.scala (hunk 3 above)
Comment (manolama) [reply atlas-cloudwatch/src/main/scala/com/netflix/atlas/cloudwatch/PublishQueue.scala:119]: The docs say Akka won't retry a `POST` though. And I do like having the counter to see if anything is retrying. Is it OK to keep this with just the backoff?

File: atlas-cloudwatch/src/main/scala/com/netflix/atlas/poller/Messages.scala (hunk 6 above)
Comment (manolama) [reply atlas-cloudwatch/src/main/scala/com/netflix/atlas/poller/Messages.scala:47]: The queue grouping returns a `Seq` instead of a `List`.

File: atlas-cloudwatch/src/test/scala/com/netflix/atlas/cloudwatch/PublishQueueSuite.scala (hunk 8 above)
Comment (manolama) [reply atlas-cloudwatch/src/test/scala/com/netflix/atlas/cloudwatch/PublishQueueSuite.scala:230]: Hmm, nice.

File: atlas-cloudwatch/src/test/scala/com/netflix/atlas/cloudwatch/PublishQueueSuite.scala (hunk 7 above)
Comment (manolama) [reply atlas-cloudwatch/src/test/scala/com/netflix/atlas/cloudwatch/PublishQueueSuite.scala:82]: Agreed, I hate it but still haven't dug into the Akka framework to figure out how to tick through the dispatcher to make sure the futures are executed.
I could refactor it for now to return a future privately.

File: atlas-cloudwatch/src/main/scala/com/netflix/atlas/cloudwatch/PublishQueue.scala (hunk 5 above)
Comment (manolama) [reply atlas-cloudwatch/src/main/scala/com/netflix/atlas/cloudwatch/PublishQueue.scala:128]: Ack, I can drop it. I like having the `stack` tag to correlate with the dropped metrics.

File: atlas-cloudwatch/src/main/scala/com/netflix/atlas/cloudwatch/PublishQueue.scala [hunk 9]
```diff
@@ -104,33 +102,44 @@ class PublishQueue(
...
           case 400 => // Bad message, all data dropped
             val id = datapointsDropped.withTag("reason", "completeFailure")
-            incrementFailureCount(id, response, size)
-          case 429 | 504 => // backoff
+            incrementFailureCount(id, response, size, promise)
+          case 429 => // backoff
             retry(payload, retries, size)
-          case v => // Unexpected, assume all dropped
             response.discardEntityBytes()
+            promise.complete(Try(null))
```
Comment (brharrington) [inline atlas-cloudwatch/src/main/scala/com/netflix/atlas/cloudwatch/PublishQueue.scala:121]: There is a [NotUsed](https://doc.akka.io/api/akka/current/akka/NotUsed.html) type in Akka that is typically use for this sort of response where the data value is not used.

**Review by brharrington — APPROVED** [review]
